benchmark/messages.cpp

```diff
@@ -0,0 +1,111 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024 Cisco Systems
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include <quicr/detail/uintvar.h>
+
+#include <benchmark/benchmark.h>
+
+#include <quicr/detail/messages.h>
+#include <quicr/detail/serializer.h>
+
+static void
+Moq_EncodeGroupHeader(benchmark::State& state)
+{
+    for ([[maybe_unused]] const auto& _ : state) {
+        quicr::Serializer buffer;
+        auto hdr = quicr::messages::MoqStreamHeaderGroup{};
+        hdr.subscribe_id = 0x100;
+        hdr.track_alias = 0x100;
+        hdr.priority = 0xA;
+        hdr.group_id = 0x1;
+        buffer << hdr;
+        benchmark::DoNotOptimize(hdr);
+        benchmark::DoNotOptimize(buffer);
+        benchmark::ClobberMemory();
+    }
+}
+
+static void
+Moq_EncodeGroupObject1000Bytes(benchmark::State& state)
+{
+    const auto payload = std::vector<uint8_t>(1000, 0);
+    for ([[maybe_unused]] const auto& _ : state) {
+        quicr::Serializer buffer;
+        auto obj = quicr::messages::MoqStreamGroupObject{};
+        obj.object_id = 0x1;
+        obj.payload.reserve(payload.size());
+        obj.payload.assign(payload.begin(), payload.end());
+        buffer << obj;
+        benchmark::DoNotOptimize(obj);
+        benchmark::DoNotOptimize(buffer);
+        benchmark::ClobberMemory();
+    }
+}
+
+static void
+Moq_EncodeGroupObject1Byte(benchmark::State& state)
+{
+    for ([[maybe_unused]] const auto& _ : state) {
+        quicr::Serializer buffer;
+        auto obj = quicr::messages::MoqStreamGroupObject{};
+        obj.object_id = 0x1;
+        obj.payload = { 0x01 };
+        buffer << obj;
+        benchmark::DoNotOptimize(obj);
+        benchmark::DoNotOptimize(buffer);
+        benchmark::ClobberMemory();
+    }
+}
+
+static void
+Moq_DecodeGroupHeader(benchmark::State& state)
+{
+    quicr::Serializer buffer;
+    auto hdr = quicr::messages::MoqStreamHeaderGroup{};
+    hdr.subscribe_id = 0x100;
+    hdr.track_alias = 0x100;
+    hdr.priority = 0xA;
+    hdr.group_id = 0x1;
+    buffer << hdr;
+
+    const auto net_data = buffer.Take();
+    quicr::StreamBuffer<uint8_t> sbuf;
+
+    for ([[maybe_unused]] const auto& _ : state) {
+        sbuf.PushLengthBytes(net_data);
+        hdr = quicr::messages::MoqStreamHeaderGroup{};
+        sbuf >> hdr;
```

Comment:
It is awkward writing benchmarks for methods that pop from a queue, but this is meauring a lot more than `sbuf >> hdr;`. There is a way to pause and resume the state timer, but that can cause some unexpected large measurement numbers. Maybe worth allocating the buffer with enough bytes for the bench that you won't have to keep pushing bytes.